keps/sig-api-machinery/2211-omit-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,372 @@
+bandwidth usage compared to simply returning the original object, and every client
+(such as kubectl or CLI tools written by users using client-go) needs to drop the managedFields by themselves.
+
+With this KEP, the API server could return objects without managedFields on demand, so that least client-side changes is needed
+to make the response less verbose. 
+
+### Goals
+
+* Omit managedFields from responses, when requested by clients.
+
+### Non-Goals
+
+* Turn off managedFields. Server-side Apply relies on managedFields, we could not simply disable it.
+* Clear manageFields only on the client-side.
+
+## Proposal
+
+### Server-side changes
+
+* Add three content types:
+  * `application/json+managed.omit`
+  * `application/yaml+managed.omit`
+  * `application/vnd.kubernetes.protobuf+managed.omit`
+* Add a new serializer, say `UnmanagedSerializer`, which wraps existing serializers and omits the managedFields before encoding objects.
+* Make the API server to use `UnmanagedSerializer` when clients properly set the `Accept` header and the selected
+  content type after negotiation is one of the above three.
+
+In addition, it is worth noting that we could add similar content types to represent specific format version of managedFields,
+such as `application/json+managed.v1`, `application/yaml+managed.v2` etc, so that users are able to request a certain
+version of managedFields with these headers.
+
+### Client-side changes
+
+Implement a flag in `kubectl` to let users choose to omit the managedFields.
+
+### Risks and Mitigations
+
+Risk:
+* The API server might not be up-to-date and cannot recognize the new content types.
+
+Mitigations:
+1. We could add the original content type to the `Accept` header as a fallback, such as
+  `Accept: application/json+managed.omit, application/json;q=0.9`, therefore the API server could return JSON-formatted 
+  objects rather than reject the request with a 406(Not Acceptable) error if it is too old to recognize the new content type.
+2. Then make `kubectl get` to drop managedFields if user wants to omit managedFields, but the API server doesn't realize that.
+
+## Design Details
+
+### Server-side
+
+1. Add a content type suffix to `k8s.io/apimachinery/pkg/runtime/types.go`, which can be used to construct the new content type 
+along with the existing ones:
+```go
+const ContentTypeSuffixOmitManagedFields = "+managed.omit"
+
+const (
+    ContentTypeJSON     string = "application/json"
+    ContentTypeYAML     string = "application/yaml"
+    ContentTypeProtobuf string = "application/vnd.kubernetes.protobuf"
+)
+```
+
+2. Add a new serializer to `k8s.io/apimachinery/pkg/runtime/serializer/` which wraps the existing serializers and
+omits the managedFields before encoding:
+```go
+type serializer struct {
+    inner      runtime.Serializer
+}
+
+func NewSerializer(s runtime.Serializer) runtime.Serializer {
```

Comment:
This name seems wrong, or we need to include the package name if that's how you want to disambiguate.